### ISSUE TITLE
[D9][SR] D7 crossport: User module's flood controls should do better logging

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5035]
This is a cross-port of https://www.drupal.org/project/drupal/issues/2989985 (merged into Drupal 7.x about 5 months ago with https://git.drupalcode.org/project/drupal/commit/51d6e29), which itself is a backport of what went into 9.1.x with https://www.drupal.org/project/drupal/issues/2983395

> ## Problem/Motivation
> At present if an attempt to authenticate via the user login form is blocked by flood control, nothing useful is logged.
> 
> The form gets an error set on it, but the response is still a 200 which means that access log entries cannot be identified by anyone interested in whether a site is being brute-forced.
> 
> e.g. `modules/user/user.module` (from `user_login_authenticate_validate()` and `user_login_final_validate())`
>
> ```php
>     if (!flood_is_allowed('failed_login_attempt_ip',
>         variable_get('user_failed_login_ip_limit', 50),
>         variable_get('user_failed_login_ip_window', 3600))) {
>       $form_state['flood_control_triggered'] = 'ip';
>       return;
>     }
> ```

> ...and then:
> 
> ```php
>     if (isset($form_state['flood_control_triggered'])) {
>       if ($form_state['flood_control_triggered'] == 'user') {
>         form_set_error('name', format_plural(variable_get('user_failed_login_user_limit', 5), 'Sorry, there has been more than one failed login attempt for this account. It is temporarily blocked. Try again later or <a href="@url">request a new password</a>.', 'Sorry, there have been more than @count failed login attempts for this account. It is temporarily blocked. Try again later or <a href="@url">request a new password</a>.', array('@url' => url('user/password'))));
>       }
>       else {
>         // We did not find a uid, so the limit is IP-based.
>         form_set_error('name', t('Sorry, too many failed login attempts from your IP address. This IP address is temporarily blocked. Try again later or <a href="@url">request a new password</a>.', array('@url' => url('user/password'))));
>       }
>     }
> ```
> 
> Similarly like in D8, we always get only a form error that is not informative as it returns with a 200.

> At present we see something like this when the login form is brute-forced:
> 
> ```
> ==> syslog <==
> Aug  2 12:47:29 xenial-php drupal: http://drupal7x.xp|1533214049|user|10.0.3.1|http://drupal7x.xp/user|http://drupal7x.xp/user|0||Login attempt failed for foo.
> 
> ==> access.log <==
> drupal7x.xp 10.0.3.1 - - [02/Aug/2018:12:47:29 +0000] "POST /user HTTP/1.1" 200 2308
> ```
> 
> ..repeated a few times, until flood control kicks in, at which point we just see:
> 
> ```
> ==> access.log <==
> drupal7x.xp 10.0.3.1 - - [02/Aug/2018:12:47:39 +0000] "POST /user HTTP/1.1" 200 
> drupal7x.xp 10.0.3.1 - - [02/Aug/2018:12:47:47 +0000] "POST /user HTTP/1.1" 200 
> drupal7x.xp 10.0.3.1 - - [02/Aug/2018:12:53:52 +0000] "POST /user HTTP/1.1" 200 
> ```

> ## Proposed resolution
> The referenced D8 issue introduces logging solution for the above cases that in D7 should be implemented by calling watchdog.

**Comment by klonos** [comment]
PR up for review: https://github.com/backdrop/backdrop/pull/3585

The only difference with the D7 backport from D9 is that in Backdrop we have a dedicated form for flood-related settings under `/admin/config/people/login`, so I've added a checkbox there, instead of a setting docblock in `settings.php`.

**Comment by klonos** [comment]
We can copy-paste the change record from the respective one for D7: https://www.drupal.org/node/3179936

**Comment by indigoxela** [comment]
Don't forget to re-open your PR :smile:

**Comment by klonos** [comment]
Sorry, boring day work got in the way 😅 ...ZenCI is acting up again 😓

**Comment by klonos** [comment]
Draft change record created here: https://docs.backdropcms.org/change-records/the-user-modules-flood-control-provides-better-logging-and-hook_user_flood_control

**Comment by indigoxela** [comment]
This is how I tested:

- created a small php script using curl to do the flooding
- logged in as admin
- reduced the flood limit for easier testing
- fired the script several times (from a different IP :smirk:)
- checked the logs

I played with different variants/combination of settings, added a user with the name, flooded again and checked the logs. Disabled flood logging, flooded again, checked the logs (no more logging).
This works for me. :+1: 

Some screenshots:

Form item:
![flood-log-form-item](https://user-images.githubusercontent.com/842066/124279563-f4229480-db47-11eb-86fe-261aa5458671.png)

Log entries:
![dblog-flood-log](https://user-images.githubusercontent.com/842066/124279598-ff75c000-db47-11eb-934f-507e105465ae.png)

**Comment by indigoxela** [comment]
One thing I'm really not sure of is: should the default really be "on"? If it were "off", then nothing would change compared to previous behavior. Personally, I'm leaning towards "off by default".

**Comment by herbdool** [comment]
RTBC. Thanks @klonos 

I've fixed the conflict. I think this is important enough to enable by default, since it would be hard to discover otherwise. And it's a good compromise to show them to disable via the UI.

**Comment by klonos** [comment]
Thank you @herbdool 🙏🏼  ...we now need to decide:
- this is a new feature: in that case, since the feature freeze for 1.21.x has passed, we'll need to set the milestone to 1.22.0
- this is a task (crossport of something that was added in D7): in that case, the 1.21.1 milestone is OK, but the type of this issue should be changed to `type - task`

**Comment by klonos** [comment]
...we also need to:
- add `@since` in newly-introduced functions
- [ ] update the "Introduced in version" field in the change record before publishing it

**Comment by herbdool** [comment]
@klonos I'm leaning towards making it a task. It's fairly small change and it's rare that a d7 crossport ends up being in the minor release. If I recall correctly.

**Comment by klonos** [comment]
Thanks @herbdool 👍🏼 ...I've updated the version number on the change record, and also added `@since` in all functions in the PR: [backdrop/backdrop@`b1c11e1` (#3585)](https://github.com/backdrop/backdrop/pull/3585/commits/b1c11e1fec3d9a978e0cf144619cc38669fe8562)

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/3585 into 1.x and 1.21.x. I wasn't sure about holding off on this until the next minor release either, but like @herbdool says it's uncommon to hold off cross ports until minor versions. So as a task and crossport putting this in 1.21.2 makes sense to me.

Thanks @klonos, @indigoxela, and @herbdool! Great work putting this together, including the naming, moving of options, the upgrade path, and test coverage!
